src/Compatibility/Microsoft.DotNet.ApiCompat.Shared/ValidateAssemblies.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using Jab;
+using Microsoft.DotNet.ApiCompatibility;
+using Microsoft.DotNet.ApiCompatibility.Abstractions;
+using Microsoft.DotNet.ApiCompatibility.Logging;
+using Microsoft.DotNet.ApiCompatibility.Runner;
+
+namespace Microsoft.DotNet.ApiCompat
+{
+    [ServiceProvider]
+    [Singleton(typeof(ISuppressionEngine), Factory = nameof(SuppressionEngineFactory))]
+    [Singleton(typeof(ICompatibilityLogger), Factory = nameof(LogFactory))]
+    [Singleton(typeof(IApiComparerFactory), typeof(ApiComparerFactory))]
+    [Singleton(typeof(IAssemblySymbolLoaderFactory), typeof(AssemblySymbolLoaderFactory))]
+    [Singleton(typeof(IMetadataStreamProvider), typeof(MetadataStreamProvider))]
+    [Singleton(typeof(IApiCompatRunner), typeof(ApiCompatRunner))]
+    internal partial class ValidateAssembliesServiceProvider
+    {
+        public Func<ICompatibilityLogger> LogFactory { get; }
+
+        public Func<ISuppressionEngine> SuppressionEngineFactory { get; }
+
+        public ValidateAssembliesServiceProvider(Func<ISuppressionEngine, ICompatibilityLogger> logFactory,
+            Func<ISuppressionEngine> suppressionEngineFactory)
+        {
+            // It's important to use GetService<T> here instead of directly invoking the factory
+            // to avoid two suppression engine being created.
+            LogFactory = () => logFactory(GetService<ISuppressionEngine>());
+            SuppressionEngineFactory = suppressionEngineFactory;
+        }
+    }
+
+    internal static class ValidateAssemblies
+    {
+        public static void Run(Func<ISuppressionEngine, ICompatibilityLogger> logFactory,
+            bool generateSuppressionFile,
+            string? suppressionFile,
+            string? noWarn, 
+            string[] leftAssemblies,
+            string[] rightAssemblies,
+            bool enableStrictMode,
+            string[][]? leftAssembliesReferences,
+            string[][]? rightAssembliesReferences,
+            bool createWorkItemPerAssembly,
+            (string CaptureGroupPattern, string ReplacementString)[]? leftAssembliesTransformationPatterns,
+            (string CaptureGroupPattern, string ReplacementString)[]? rightAssembliesTransformationPatterns)
+        {
+            // Configure the suppression engine. Ignore the passed in suppression file if it should be generated and doesn't yet exist.
+            string? suppressionFileForEngine = generateSuppressionFile && !File.Exists(suppressionFile) ? null : suppressionFile;
+            Func<ISuppressionEngine> suppressionEngineFactory = () => new SuppressionEngine(suppressionFileForEngine, noWarn, generateSuppressionFile);
```

Comment:
>suppressionEngineFactory [](http://example.com/codeflow?start=37&length=24)

I'm not sure I understand why this isn't fetched via the serviceProvider like the rest.

Reply:
Oh got it, I see where this is used.